### PR TITLE
Ring algorithm for reduce

### PR DESCRIPTION
**Pull request by akhillanger**
1. Adds CVAR for chunk size specification in Tree based reduce
2. Add ring algorithm for reduce

### CODE REVIEW

**Comment by akhillanger** [comment]
This pull request is based on top of #3020.

**Comment by akhillanger** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by akhillanger** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by akhillanger** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by akhillanger** [comment]
@raffenet I have addressed your comments.

File: src/mpi/coll/ireduce/ireduce_tsp_algos.h [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+    num_children = my_tree.num_children;
+
+    /* identify my locaion in the tree */
+    is_tree_root = (rank == tree_root) ? 1 : 0;
+    is_tree_leaf = (num_children == 0 && !is_tree_root) ? 1 : 0;
+    is_tree_intermediate = (!is_tree_leaf && !is_tree_root);
+
+    /* allocate buffers to receive data from children */
+    if (!is_tree_leaf) {
+        child_buffer = MPIR_TSP_sched_malloc(sizeof(void *) * num_children, sched);
```
[the remark this answers is not in the document]
Comment (raffenet) [reply src/mpi/coll/ireduce/ireduce_tsp_algos.h]: OK, somehow I missed the reduce buffers being allocated below, so I see its used for more than one type of buffer. I think an improved commit message + comment on the addition of that API could help explain its use. From my understanding, it is for allocating temporary communication buffers used by operations in a schedule. Is that correct?

[the remark this answers is not in the document]
Comment (akhillanger) [reply src/mpi/coll/ireduce/ireduce_tsp_algos.h]: Yes, your understanding is correct. I will update the commit message for that API addition commit.

**Comment by raffenet** [comment]
Did the scope of this PR change? There's a bunch more commits than there were yesterday.

**Comment by raffenet** [comment]
Hmm, I seem to have confused this with #3020. I'm going to move back to that one...

**Comment by akhillanger** [comment]
It is same as yesterday. It has 2 unrelated (to the title of the PR) commits
`coll: Add CVAR for chunk size specification in ireduce`
`test/coll: Add tests for tree based pipelining in reduce algos`

This PR is based on #3020, which has 7 of the 12 commits in this PR.

**Comment by raffenet** [comment]
@akhillanger can you rebase this on the commit merged this morning that fixes the testsuite?

**Comment by akhillanger** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by akhillanger** [comment]
@raffenet Done.

**Comment by raffenet** [comment]
Looks like this produces bad results with the UCX netmod. https://jenkins.mpich.org/job/mpich-review-ch4-ucx/1201/. Can you investigate?

**Comment by akhillanger** [comment]
Sure. I will take a look. Meanwhile you can continue the review.

File: src/mpi/coll/ibcast/ibcast_tsp_ring_algos.h [hunk 2]
```diff
@@ -59,6 +59,10 @@ int MPIR_TSP_Ibcast_sched_intra_ring(void *buffer, int count, MPI_Datatype datat
                                              maxbytes, count, num_chunks,
                                              chunk_size_floor, chunk_size_ceil));
 
+    /* For correctness, transport based collectives need to get the
+     * tag from the same pool as schedule based collectives */
+    MPIDU_Sched_next_tag(comm, &tag);
```
Comment (raffenet) [inline src/mpi/coll/ibcast/ibcast_tsp_ring_algos.h]: We just had a PR that converted these to `MPIR`.

File: src/mpi/coll/ibcast/ibcast_tsp_ring_algos.h [hunk 3]
```diff
@@ -59,6 +59,10 @@ int MPIR_TSP_Ibcast_sched_intra_ring(void *buffer, int count, MPI_Datatype datat
                                              maxbytes, count, num_chunks,
                                              chunk_size_floor, chunk_size_ceil));
 
+    /* For correctness, transport based collectives need to get the
```
Comment (raffenet) [inline src/mpi/coll/ibcast/ibcast_tsp_ring_algos.h]: This would be helpful to explain in the commit message.

File: src/mpi/coll/ireduce/ireduce.c [hunk 4]
```diff
@@ -20,6 +20,18 @@
       description : >-
         k value for tree based ireduce (for tree_kary and tree_knomial)
 
+    - name        : MPIR_CVAR_IREDUCE_TREE_PIPELINE_CHUNK_SIZE
```
Comment (raffenet) [inline src/mpi/coll/ireduce/ireduce.c]: This has nothing to do with the ring algorithm, right?

File: test/mpi/coll/test_coll_algos.sh [hunk 5]
```diff
@@ -45,7 +45,7 @@ for algo_name in ${algo_names}; do
     for kval in ${kvalues}; do
         #set the environment
         env="${testing_env} env=MPIR_CVAR_IREDUCE_INTRA_ALGORITHM=${algo_name} "
-        env+="env=MPIR_CVAR_IREDUCE_TREE_KVAL=${kval} "
+        env+="env=MPIR_CVAR_IREDUCE_TREE_KVAL=${kval} env=MPIR_CVAR_IREDUCE_TREE_PIPELINE_CHUNK_SIZE=4096 "
```
Comment (raffenet) [inline test/mpi/coll/test_coll_algos.sh]: Also seems out of place in this PR.

File: src/mpi/coll/ibcast/ibcast_tsp_ring_algos.h (hunk 2 above)
Comment (akhillanger) [reply src/mpi/coll/ibcast/ibcast_tsp_ring_algos.h]: Thanks, I will fix this here as well.

File: src/mpi/coll/ibcast/ibcast_tsp_ring_algos.h (hunk 3 above)
Comment (akhillanger) [reply src/mpi/coll/ibcast/ibcast_tsp_ring_algos.h]: This is true for all the algorithms that we are adding - they use tags from the same pool as the older schedule based collectives. This is required for correctness.

File: src/mpi/coll/ireduce/ireduce.c (hunk 4 above)
Comment (akhillanger) [reply src/mpi/coll/ireduce/ireduce.c]: In addition to ring based algo for reduce, this PR has two additional commits that enable pipelining in the already existing tree algo for reduce. See first comment in this PR. I should update the PR title.

File: test/mpi/coll/test_coll_algos.sh (hunk 5 above)
Comment (akhillanger) [reply test/mpi/coll/test_coll_algos.sh]: Same reason as above.

File: src/mpi/coll/ibcast/ibcast_tsp_ring_algos.h (hunk 2 above)
Comment (akhillanger) [reply src/mpi/coll/ibcast/ibcast_tsp_ring_algos.h]: Actually, the rebase will fix this. Let me rebase and push.

**Comment by akhillanger** [comment]
test:jenkins/ch4/ucx

File: src/mpi/coll/ireduce/ireduce.c (hunk 4 above)
Comment (akhillanger) [reply src/mpi/coll/ireduce/ireduce.c]: For completeness, in addition, it has one more commit - `Use separate tags for each task`. This is the only PR with more than 1 subtopic commit, PRs after this are consistent.

File: src/mpi/coll/ibcast/ibcast_tsp_ring_algos.h (hunk 3 above)
Comment (raffenet) [reply src/mpi/coll/ibcast/ibcast_tsp_ring_algos.h]: This commit is still missing a commit message. Please add one.

Comment (akhillanger) [reply src/mpi/coll/ibcast/ibcast_tsp_ring_algos.h]: Oh Ok. Sure.

File: src/mpi/coll/ibcast/ibcast_tsp_ring_algos.h [hunk 6]
```diff
@@ -22,8 +22,7 @@
 #undef FCNAME
 #define FCNAME MPL_QUOTE(FUNCNAME)
 int MPIR_TSP_Ibcast_sched_intra_ring(void *buffer, int count, MPI_Datatype datatype, int root,
-                                     int tag, MPIR_Comm * comm, int maxbytes,
-                                     MPIR_TSP_sched_t * sched)
+                                     MPIR_Comm * comm, int maxbytes, MPIR_TSP_sched_t * sched)
```
Comment (raffenet) [inline src/mpi/coll/ibcast/ibcast_tsp_ring_algos.h]: Unrelated whitespace?

Comment (akhillanger) [reply src/mpi/coll/ibcast/ibcast_tsp_ring_algos.h]: No, this is a consequence of removing `tag` argument and then the style-checker asking to collate the lines.

File: src/mpi/coll/ibcast/ibcast_tsp_ring_algos.h (hunk 3 above)
Comment (akhillanger) [reply src/mpi/coll/ibcast/ibcast_tsp_ring_algos.h]: Here is my commit msg

`coll: Use separate tags for each task

    So far we were using a single tag for the entire collective
    schedule. But that will not work for all the collective
    operations. A collective operation may need more than one
    tag for its operations to prevent any message mismatches.
    Therefore, tag should be assgined and stored with each task in the
    schedule.`

I will push it once the running tests complete.

**Comment by akhillanger** [comment]
@raffenet It seems that my fix resolved the ucx failures. https://jenkins.mpich.org/job/mpich-review-ch4-ucx/1202/
I will push (wth the commit msg update) once you have seen this?

**Comment by akhillanger** [comment]
The ch4/ucx test has completed for a while now. Does it take a lot of time to get the status updated below?

**Comment by raffenet** [comment]
Unclear to me why the status hasn't updated. My guess is something similar to why webhooks don't always immediately trigger tests.

**Comment by akhillanger** [comment]
Ok. Should I push now? One of the commits has an updated commit message.

**Comment by raffenet** [comment]
yes, go ahead. i will consider these tests passed.

File: src/mpi/coll/ibcast/ibcast_tsp_tree_algos.h [hunk 7]
```diff
@@ -67,6 +68,10 @@ int MPIR_TSP_Ibcast_sched_intra_tree(void *buffer, int count, MPI_Datatype datat
         MPIR_ERR_POP(mpi_errno);
     num_children = my_tree.num_children;
 
+    /* For correctness, transport based collectives need to get the
+     * tag from the same pool as schedule based collectives */
+    MPIDU_Sched_next_tag(comm, &tag);
```
Comment (raffenet) [inline src/mpi/coll/ibcast/ibcast_tsp_tree_algos.h]: MPIDU->MPIR

Comment (akhillanger) [reply src/mpi/coll/ibcast/ibcast_tsp_tree_algos.h]: Oops. Looks like I missed the fixup patch. Fixed now.
